### ISSUE TITLE
Sync access randomly fails with Error from socket

### ISSUE THREAD

**Issue by ggardet** [codetheweb/tuyapi#326]
**Describe the bug**
While using v5.3.1 (with sync access) to control a konyks polyco smartstrip (https://konyks.com/produit/polyco/), I randomly get the following error:
```
Error: Error from socket
    at Socket.<anonymous> (/var/lib/openqa/share/tests/opensuse/data/generalhw_scripts/node_modules/tuyapi/index.js:350:30)
    at Socket.emit (events.js:311:20)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
Emitted 'error' event on TuyaDevice instance at:
    at Socket.<anonymous> (/var/lib/openqa/share/tests/opensuse/data/generalhw_scripts/node_modules/tuyapi/index.js:350:16)
    at Socket.emit (events.js:311:20)
    [... lines matching original stack trace ...]
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
```

**To Reproduce**
Get state, set state and get state again, synchronously.

**Expected behavior**
It should not throw an exception.

**Debug Output**
None.

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: [e.g. macOS] openSUSE Tumbleweed
 - OS Version [e.g. 22] (rolling release)
 - Node Version [output of `node -v`] v12.16.1

**Additional context**

**Comment by codetheweb** [comment]
Can you please provide the script you're using with this?

**Comment by ggardet** [comment]
Sure, here it is:
```
const TuyAPI = require('tuyapi');

const device = new TuyAPI({
  id: process.argv[2],
  key: process.argv[3]});

const index = process.argv[4] || 1;
if (process.argv[5] === "on") {
  state = true;
}
else {
  state = false;
}

(async () => {
  await device.find();
  await device.connect();

  // Get current status (optional)
  let status = await device.get({dps: index});
  console.log(`Current status of plug#${index}: ${status}.`);

  // Set wanted value and check
  await device.set({set: state, dps: index});
  status = await device.get({dps: index});
  console.log(`New status of plug#${index}: ${status}.`);

  device.disconnect();
})();
```

**Comment by codetheweb** [comment]
That seems straightforward enough not to have any bugs.

The socket error is most likely caused by either another device trying to control it at the same time (possibly the app on your phone, depending on how you set it up) or TuyAPI is sending a malformed packet.

Could you try running your script with the environment variable `DEBUG` set to `*` (i.e. `DEBUG=* node ...`) and try to reproduce the issue?

**Comment by ggardet** [comment]
Here is the log:
```
2020-06-05T10:21:39.212Z TuyAPI Finding missing IP undefined or ID XXXobfuscatedXXX
2020-06-05T10:21:43.841Z TuyAPI Received UDP message.
2020-06-05T10:21:43.845Z TuyAPI UDP data:
2020-06-05T10:21:43.848Z TuyAPI {
  payload: {
    ip: '192.168.0.101',
    gwId: 'XXXobfuscatedXXX',
    active: 2,
    ability: 0,
    mode: 0,
    encrypt: true,
    productKey: 'u1tniiwet1pstec9',
    version: '3.3'
  },
  leftover: false,
  commandByte: 19,
  sequenceN: 0
}
2020-06-05T10:21:43.850Z TuyAPI Connecting to 192.168.0.101...
2020-06-05T10:21:43.856Z TuyAPI Socket connected.
2020-06-05T10:21:43.857Z TuyAPI GET Payload:
2020-06-05T10:21:43.857Z TuyAPI {
  gwId: 'XXXobfuscatedXXX',
  devId: 'XXXobfuscatedXXX',
  t: '1591352504',
  dps: {},
  uid: 'XXXobfuscatedXXX'
}
2020-06-05T10:21:43.861Z TuyAPI GET Payload:
2020-06-05T10:21:43.861Z TuyAPI {
  gwId: 'XXXobfuscatedXXX',
  devId: 'XXXobfuscatedXXX',
  t: '1591352504',
  dps: {},
  uid: 'XXXobfuscatedXXX'
}
2020-06-05T10:21:43.864Z TuyAPI Error event from socket. 192.168.0.101 Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:205:27) {
  errno: 'ECONNRESET',
  code: 'ECONNRESET',
  syscall: 'read'
}
events.js:288
      throw er; // Unhandled 'error' event
      ^

Error: Error from socket
    at Socket.<anonymous> (/var/lib/openqa/share/tests/opensuse/data/generalhw_scripts/node_modules/tuyapi/index.js:350:30)
    at Socket.emit (events.js:311:20)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
Emitted 'error' event on TuyaDevice instance at:
    at Socket.<anonymous> (/var/lib/openqa/share/tests/opensuse/data/generalhw_scripts/node_modules/tuyapi/index.js:350:16)
    at Socket.emit (events.js:311:20)
    [... lines matching original stack trace ...]
    at processTicksAndRejections (internal/process/task_queues.js:84:21) at /usr/lib/os-autoinst/backend/generalhw.pm line 83.
```

**Comment by codetheweb** [comment]
Please try commenting out the `GET` on line 403 of index.js and see if that does anything:
https://github.com/codetheweb/tuyapi/blob/cea119ecdd476fe77364394df18448783cfd454a/index.js#L403

If I remember correctly, other users have had issues with that in the past. If that fixes it, we can look at a more permanent solution.

**Comment by ggardet** [comment]
In my case it did not help.

**Comment by codetheweb** [comment]
@ggardet can you try using the [event based API](https://github.com/codetheweb/tuyapi#asynchronous-event-based-recommended) and see if that changes anything?

@tgeorge91 are you using the callback/Promise or event API?

**Comment by ggardet** [comment]
In async mode, it seems to be more stable (to be confirmed), but I sometimes get:
```
Error! Error: Error from socket
    at Socket.<anonymous> (/var/lib/openqa/share/tests/opensuse/data/generalhw_scripts/node_modules/tuyapi/index.js:350:30)
    at Socket.emit (events.js:311:20)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
```

**Comment by codetheweb** [comment]
@ggardet there already is some basic retry logic: https://github.com/codetheweb/tuyapi/blob/cea119ecdd476fe77364394df18448783cfd454a/index.js#L238

If you want to implement more, PRs are welcome. 😉 

Can you also run your script with `DEBUG=*` again to get the full log (or is it the exact same error)?

**Comment by ggardet** [comment]
Here is the log:
```
  TuyAPI Finding missing IP undefined or ID XXX +0ms
  TuyAPI Received UDP message. +1s
  TuyAPI UDP data: +4ms
  TuyAPI {
  TuyAPI   payload: {
  TuyAPI     ip: '192.168.0.101',
  TuyAPI     gwId: 'XXX',
  TuyAPI     active: 2,
  TuyAPI     ability: 0,
  TuyAPI     mode: 0,
  TuyAPI     encrypt: true,
  TuyAPI     productKey: 'XXX',
  TuyAPI     version: '3.3'
  TuyAPI   },
  TuyAPI   leftover: false,
  TuyAPI   commandByte: 19,
  TuyAPI   sequenceN: 0
  TuyAPI } +0ms
  TuyAPI Connecting to 192.168.0.101... +6ms
  TuyAPI Socket connected. +12ms
Connected to device!
  TuyAPI Error event from socket. 192.168.0.101 Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:205:27) {
  errno: 'ECONNRESET',
  code: 'ECONNRESET',
  syscall: 'read'
} +7ms
Error! Error: Error from socket
    at Socket.<anonymous> (/var/lib/openqa/share/tests/opensuse/data/generalhw_scripts/node_modules/tuyapi/index.js:350:30)
    at Socket.emit (events.js:311:20)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
  TuyAPI Socket closed: 192.168.0.101 +3ms
Disconnected from device.
  TuyAPI Disconnect +9s
```

**Comment by codetheweb** [comment]
Thanks.

Before it was throwing an error when getting/setting state, correct?

Looks like it's throwing an error now when initially connecting to the device.

**Comment by ggardet** [comment]
It looks like the connection error can happen at any time: initial connection, a get or a set.

**Comment by codetheweb** [comment]
Ok. If it's throwing connection errors randomly, the most likely cause is something else trying to connect to it / control it.
It's hard to debug further than that because the issue could be due to a multitude of factors, including flaky device hardware and weird network setups.

**Comment by ggardet** [comment]
Could we mitigate this with a retry mechanism, which could be optional maybe?

**Comment by kueblc** [comment]
There is already a retry mechanism

**Comment by ggardet** [comment]
> There is already a retry mechanism

Not for this specific issue apparently.

**Comment by codetheweb** [comment]
Implementing a retry mechanism for this would require a large rewrite to how we handle device state. It's currently 'unmanaged', meaning that TuyAPI doesn't know or care what the device state is - that's left up to the user. If device state was instead 'managed', meaning TuyAPI knew with 99% certainty what state the device is / should be in, then it would be possible to implement retry for set.

I was planing to do something like this for [v6](https://github.com/codetheweb/tuyapi/issues/169): @tuyapi/driver would be a very basic, 'dumb' package. Then @tuyapi/devices could be written on top of that with managed device state.

**TL;DR**: not doable currently, maybe in a future version.

**Comment by ggardet** [comment]
As a workaround, I have a retry mechanism in a wrapper script.

**Comment by geotrieu** [comment]
@ggardet Would you be able to post the wrapped here? I am also having trouble with this, it is constantly throwing the` Error from socket `error

**Comment by ggardet** [comment]
@geotrieu, I did something like this in a bash script:
```
# Retry mechanism, so do not fail
set +e

n=0
until [ "$n" -ge 5 ]
do
   # Run JS script, which can fail and must be retried - poo#69052
   node $(dirname $0)/tuya_sync.js $tuya_ID $tuya_key $tuya_index $tuya_state \
   && break  # Break if succeeds
   n=$((n+1))
   echo "Attempt $n failed, retry"
   sleep 2
done

# Fail globally if last attempt failed
set -e
if [ "$n" -ge 5 ]; then
   exit 1;
fi
```
